ls1intum/Artemis
# Enchancement/upgrade ocaml template

Supersedes #3938 on an internal branch and with additional fixes and improvements

### Checklist
#### General


#### Server & Client
No significant changes
#### Changes affecting Programming Exercises
  OCaml is not available here

### Motivation and Context
This year the FPV course wants to use Artemis with OCaml. The current OCaml template is rather simple and is missing some features that we would like to have.

### Description
- Add support for solution repository checkout
- Add support for not running some tests during the submission period
- Show build errors in student code instead of hiding all build output
- Disallow symlinks in the submission to prevent students from just including the solution
- Changing the template to make it easier for students to test and run their code with `dune utop src`
- Change tests to be based around qcheck
- Support restricting students to a subset of ocaml (mainly functional code only)
- Support for restricting use of and mocking stdlib 
- removed batteries and core libraries as they are not relevant for teaching ocaml
- changed the docker image to https://github.com/bottbenj/docker-test/pkgs/container/docker-test as the proper image is still outdated and even more changes have become necessary.

### Steps for Testing

Prerequisites:
- 1 Instructor
1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to some test course
4. Create a new OCaml exercise
    Make sure to check the enable solution checkout box
5. Wait for the builds to complete and check that both template and solution build, that the template passes 0/9 tests and that the solution passes 6/9 tests, the name of the failing tests should end with `[hidden]` and the message should be `This test is hidden`
6. Change the variables HIDDEN_START and HIDDEN_END in run.sh in the tests repo to not include the current date
7. When re-running tests before HIDDEN_START and after HIDDEN_END the solution should pass 9/9 tests and the `[hidden]` tests of the template should now fail with a TODO message.
8. Cleanup the exercise and course, if required

### Review Progress





#### Code Review
#### Manual Tests
